Lifecycle Events

kadaan opened the pull request:
Add ability to define additional JobListeners in the job properties, added new JobListener lifecycle events, and moved the listener code into it's own package

kadaan commented:
@liyinan926 Can you review this?

liyinan926 commented:
Sure, I will take a look at a later time today.

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/State.java:
```diff
@@ -244,7 +244,7 @@ public String getProp(String key, String def) {
    */
   public List<String> getPropAsList(String key, String def) {
-    return Splitter.on(",").trimResults().omitEmptyStrings().splitToList(getProperty(key, def));
+    return Splitter.on(",").trimResults().omitEmptyStrings().splitToList(getProp(key, def));
```
You might want to define a constant for this `Splitter` to avoid creating a new instance every time this method is called.

kadaan replied:
Will do it shortly.

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/State.java:
```diff
@@ -254,7 +254,19 @@ public String getProp(String key, String def) {
    */
   public Set<String> getPropAsSet(String key) {
-      return Sets.newHashSet(Splitter.on(",").trimResults().omitEmptyStrings().splitToList(getProperty(key)));
+      return Sets.newHashSet(Splitter.on(",").trimResults().omitEmptyStrings().splitToList(getProp(key)));
```
Ditto.

kadaan replied:
Will do it shortly.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -484,9 +521,26 @@ private void unlockJob() {
     jobListeners.add(jobListener);
+
+    Set<String> jobListenerClassNames = jobState.getPropAsSet(ConfigurationKeys.JOB_LISTENERS_KEY, StringUtils.EMPTY);
+    if (jobListenerClassNames != null) {
```
`jobListenerClassNames` should never be `null`.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -484,9 +521,26 @@ private void unlockJob() {
+            @SuppressWarnings("unchecked")
+            Class<? extends JobListener> jobListenerClass =
+                    (Class<? extends JobListener>) Class.forName(jobListenerClassName);
+            Constructor<? extends JobListener> jobListenerConstructor = jobListenerClass.getConstructor();
```
Why not just do `jobListenerClass.newInstance()`?

kadaan replied:
I was using a similar method to how DataPublishers where instantiated.  I can switch this to newInstance.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -484,9 +521,26 @@ private void unlockJob() {
+            jobListeners.add(jobListenerConstructor.newInstance());
+          } catch (ClassNotFoundException | NoSuchMethodException | InstantiationException |
+                  IllegalAccessException | InvocationTargetException e) {
+              LOG.warn("JobListener %s could not be created due to %s", jobListenerClassName, e.getMessage());
```
Please use `LOG.warn(message, e)`;

kadaan replied:
Okay

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -731,4 +785,23 @@ private static void cleanupStagingDataForEntireJob(JobState jobState) {
+                               JobListenerAction action)
+          throws JobException {
+    TimingEvent timer =
+        this.eventSubmitter.getTimingEvent(timerEventName);
```
Line 792 and 793 can be combined.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/listeners/JobListeners.java:
```diff
@@ -73,13 +74,55 @@ public Void call()
+          @Override
+          public Void call()
+              throws Exception {
+            jobListener.onJobCompletion(jobContext);
```
If any `jobListener` throws an exception, the remaining ones (if any) will be skipped. Is this behavior desirable?

kadaan replied:
This is how this code was before.  I agree it is not ideal.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/listeners/AbstractJobListener.java:
```diff
@@ -0,0 +1,50 @@
+ * @author Joel Baranick
+ */
+public abstract class AbstractJobListener implements JobListener {
+  @Override
```
Empty line before `@Override` please.

kadaan commented:
@liyinan926 Addressed your comments.  BTW, I think if we can really beef up the lifecycle events, that it will simplify AbstractJobLauncher and allow gobblin to have much more flexibility.

kadaan commented:
Grrrr.  Intermittently failing tests!!!!!!

kadaan commented:
@liyinan926 Addressed comments

kadaan commented:
@liyinan926 How is that?

liyinan926 commented:
@zliu41 or @sahilTakiar can you help merge this? I won't since I can't watch for the hudson builds.

sahilTakiar commented:
Thanks for taking a look @liyinan926 

Merged

kadaan commented:
@liyinan926 <https://github.com/liyinan926> @sahilTakiar Thanks for reviewing and merging this.
